Add doc for NavigationServer, NavigationAgent and RVO agent avoidance (#6049)
godotengine/godot-docs: tutorials/navigation/navigation_using_navigationagents.rst

.. _doc_navigation_using_navigationagents:

Using NavigationAgents
======================

NavigationsAgents are helper nodes to facilitate common calls to the NavigationServer API 
on behalf of the parent actor node in a more convenient manner for beginners.

2D and 3D version of NavigationAgents are available as 
:ref:`NavigationAgent2D<class_NavigationAgent2D>` and 
:ref:`NavigationAgent3D<class_NavigationAgent3D>`  respectively.

NavigationsAgents are entirely optional for navigation pathfinding. 
The functionality of NavigationsAgents can be recreated with scripts and direct 
calls to the NavigationServer API. If the default NavigationsAgent does not do what you want 
for your game feel free to design your own NavigationsAgent with scripts.

.. warning::

    NavigationsAgent nodes and NavigationServer ``agents`` are not the same. 
    The later is an RVO avoidance agent and solely used for avoidance. 
    RVO avoidance agents are not involved in regular pathfinding.

NavigationAgent Pathfinding
---------------------------

> To use NavigationAgents for pathfinding place a NavigationAgent2D/3D Node below a Node2D/3D inheriting parent Node.

Reviewer: ```suggestion
To use NavigationAgents for pathfinding, place a NavigationAgent2D/3D Node below a Node2D/3D inheriting parent node.
```

> To have the agent query a path to a target position use ``set_target_location()`` method.

Reviewer: ```suggestion
To have the agent query a path to a target position use the ``set_target_location()`` method.
```

> Once the target has been set the next position to follow in the path 

Reviewer: ```suggestion
Once the target has been set, the next position to follow in the path 
```

> can be received with the ``get_next_location()`` function. Move the parent actor node 

Reviewer: ```suggestion
can be retrieved with the ``get_next_location()`` function. Move the parent actor node 
```

> to this position with your own movement code. On the next ``physics_frame`` call 
> again ``get_next_location()`` for the next position until the path ends.

Reviewer: ```suggestion
to this position with your own movement code. On the next ``physics_frame``, call 
``get_next_location()`` again for the next position and repeat this until the path ends.
```

> NavigationAgents have there own internal logic to proceed with the current path and call for updates.

Reviewer: ```suggestion
NavigationAgents have their own internal logic to proceed with the current path and call for updates.
```

NavigationAgents recognize by distance when a path point or the final target is reached.

> NavigationAgents refresh a path automatically when to far away from the current pathpoint.

Reviewer: ```suggestion
NavigationAgents refresh a path automatically when too far away from the current pathpoint.
```

The important updates are all triggered with the ``get_next_location()`` function 
when called in ``_physics_process()``.

> Be careful calling other NavigationAgent functions not required for pathmovement 

Reviewer: ```suggestion
Be careful calling other NavigationAgent functions not required for path movement 
```

> while the actor is following a path as many function trigger a full path refresh.

Reviewer: ```suggestion
while the actor is following a path, as many function trigger a full path refresh.
```

.. note::

    New NavigationAgents will automatically join the 
    default navigation map for their 2D/3D dimension.

.. warning::

    Resetting the path every frame (by accident) might get the actor to stutter or spin around in place.

NavigationAgents were designed with ``_physics_process()`` in mind to keep in sync with both :ref:`NavigationServer3D<class_NavigationServer3D>` and :ref:`PhysicsServer3D<class_PhysicsServer3D>`.

They work well out of the box with :ref:`CharacterBody2D<class_CharacterBody2D>` and :ref:`CharacterBody3D<class_CharacterBody3D>` as well as any rigid bodies.

> The important restriction for non-physics characters is that the NavigationAgent node only accepts a single update each ``physics_frame`` as further updates will be blocked.

Reviewer: This may be worth more noticeable formatting, i.e. a warning or similar.

  - The property ``max_speed`` controls the maximum velocity assumed for the agents avoidance calculation.
    If the agents parents moves faster than this value the avoidance ``safe_velocity`` might not be accurate enough to avoid collision.

The ``velocity_computed`` signal of the agent node must be connected to receive the ``safe_velocity`` calculation result.

.. image:: img/agent_safevelocity_signal.png

Additional the current velocity of the agents parent must be set for the agent in ``_physics_process()`` with ``set_velocity()``.

After a short wait for processing the avoidance (still in the same frame) the ``safe_velocity`` vector will be received with the signal. 
This velocity vector should be used to move the NavigationAgent's parent node in order to avoidance collision with other avoidance registered agents in proximity.

RVO exists in its own space and has no information from navigation meshes or physics collision.
Behind the scene avoidance agents are just circles with different radius on a flat plane.

> In narrow places obstructed with collision objects the avoidance maneuver radius needs to be 

Reviewer: ```suggestion
In narrow places obstructed with collision objects, the avoidance maneuver radius needs to be 
```
